src/lib.rs

```diff
@@ -3,27 +3,1109 @@
+
+        <Stakes<T>>::insert(stake_id, stake);
+
+        Self::deposit_funds_into_stake_pool(value);
+
+        Ok(())
+    }
+
+    pub fn stake_from_account(
+        stake_id: &T::StakeId,
+        source_account_id: &T::AccountId,
+        value: BalanceOf<T>,
+    ) -> Result<(), StakeActionError<StakingFromAccountError>> {
+        ensure!(
+            <Stakes<T>>::exists(stake_id),
+            StakeActionError::StakeNotFound
+        );
+
+        let mut stake = Self::stakes(stake_id);
+
+        stake.start_staking(value, T::Currency::minimum_balance())?;
+
+        Self::transfer_funds_from_account_into_stake_pool(source_account_id, value)?;
+
+        <Stakes<T>>::insert(stake_id, stake);
+
+        Ok(())
+    }
+
+    /// Moves funds from specified account into the module's account
+    fn transfer_funds_from_account_into_stake_pool(
+        source: &T::AccountId,
+        value: BalanceOf<T>,
+    ) -> Result<(), TransferFromAccountError> {
+        // We don't use T::Currency::transfer() to prevent fees being incurred.
+        let negative_imbalance = T::Currency::withdraw(
+            source,
+            value,
+            WithdrawReason::Transfer,
+            ExistenceRequirement::AllowDeath,
+        )
+        .map_err(|_err| TransferFromAccountError::InsufficientBalance)?;
+
+        Self::deposit_funds_into_stake_pool(negative_imbalance);
+        Ok(())
+    }
+
+    fn deposit_funds_into_stake_pool(value: NegativeImbalance<T>) {
+        // move the negative imbalance into the stake pool
+        T::Currency::resolve_creating(&Self::stake_pool_account_id(), value);
+    }
+
+    /// Moves funds from the module's account into specified account. Should never fail if used internally.
+    /// Will panic! if value exceeds balance in the pool.
+    fn transfer_funds_from_pool_into_account(destination: &T::AccountId, value: BalanceOf<T>) {
+        let imbalance = Self::withdraw_funds_from_stake_pool(value);
+        T::Currency::resolve_creating(destination, imbalance);
+    }
+
+    /// Withdraws value from the pool and returns a NegativeImbalance.
+    /// As long as it is only called internally when executing slashes and unstaking, it
+    /// should never fail as the pool balance is always in sync with total amount at stake.
+    fn withdraw_funds_from_stake_pool(value: BalanceOf<T>) -> NegativeImbalance<T> {
+        // We don't use T::Currency::transfer() to prevent fees being incurred.
+        T::Currency::withdraw(
+            &Self::stake_pool_account_id(),
+            value,
+            WithdrawReason::Transfer,
+            ExistenceRequirement::AllowDeath,
+        )
+        .expect("pool had less than expected funds!")
+    }
+
+    /// Dry run to see if the state of stake allows for increasing stake. This should be called
+    /// to make sure increasing stake is possible before withdrawing funds.
+    pub fn ensure_can_increase_stake(
```

Comment:
Yes its genuinely public, and intended to be used  by client modules before attempting to increase stake with an imbalance to avoid having to deal with the imbalance with a failed call.

Your suggestion makes sense if its important to bubble up the reason why the call would fail.
If we don't need that information then the method can simple return a `bool`.